### PR TITLE
Add support for custom token fee - ERC20 based rollups

### PR DESCRIPTION
**Pull request by gvladika**
This commit adds support for custom token fee - new feature which enables deploying rollups which use native ERC20 token to pay for transaction and crosschain message fees instead of ETH. It means that fee token will be escrowed on L1 side in the bridge contract and it will be minted on L2 side as native currency.

Implementation is done through using inheritance to avoid code duplication. Ie. in addition to existing `Inbox` contract we have now `ERC20Inbox`. Also abstract contract `AbsInbox` is created, where all the common logic is moved to. `Inbox` and `ERC20Inbox` only contain logic that differs Eth and ERC20 based rollups. Same pattern is used for `Bridge`,  `Outbox`, `BridgeCreator`, `RollupCreator` and `RollupEventInbox`.

`ERC20Inbox` offers `depositERC20` function to deposit native fee token,  analogous to `depositEth` in `Inbox`.

One additional difference is that in ERC20-based rollup we don't charge submission fee. Reason for that is that keeping exchange rate between Eth and fee token is complex so we want to avoid it.

Test coverage is increased using Foundry unit test cases.

### CODE REVIEW

File: src/bridge/ERC20Bridge.sol [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * @title Staging ground for incoming and outgoing messages
+ * @notice Unlike the standard Eth bridge, native token bridge escrows the custom ERC20 token which is
+ * used as native currency on L2.
+ */
+contract ERC20Bridge is AbsBridge, IERC20Bridge {
+    using SafeERC20 for IERC20;
+
+    /// @inheritdoc IERC20Bridge
+    address public nativeToken;
```
Comment (PlasmaPower) [inline src/bridge/ERC20Bridge.sol:23]: We should add a storage gap field to the end of AbsBridge similar to the openzeppelin upgradeable contracts, so we can add new fields to AbsBridge without them conflicting with this field of ERC20Bridge

Comment (gvladika) [reply src/bridge/ERC20Bridge.sol:23]: done

File: src/bridge/IInbox.sol [hunk 2]
```diff
@@ -112,67 +57,6 @@ interface IInbox is IDelayedMessageProvider {
...
-     * @param l2CallValue call value for retryable L2 message
-     * @param maxSubmissionCost Max gas deducted from user's L2 balance to cover base submission fee
-     * @param excessFeeRefundAddress gasLimit x maxFeePerGas - execution cost gets credited here on L2 balance
-     * @param callValueRefundAddress l2Callvalue gets credited here on L2 if retryable txn times out or gets cancelled
-     * @param gasLimit Max gas deducted from user's L2 balance to cover L2 execution. Should not be set to 1 (magic value used to trigger the RetryableData error)
-     * @param maxFeePerGas price bid for L2 execution. Should not be set to 1 (magic value used to trigger the RetryableData error)
-     * @param data ABI encoded data of L2 message
-     * @return unique message number of the retryable transaction
-     */
-    function createRetryableTicket(
```
Comment (gzeoneth) [inline src/bridge/IInbox.sol:137]: found it a bit annoying that this changed the inbox interface, can we have IInbox as the eth inbox, and make sth like IInboxBase e.g.
IEthInbox -> IInbox
IInbox -> IInboxBase

Comment (gvladika) [reply src/bridge/IInbox.sol]: right,  makes sense to preserve IInbox compatibility. updated

File: src/bridge/ERC20Inbox.sol [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        return 0;
+    }
+
+    function _deliverToBridge(
+        uint8 kind,
+        address sender,
+        bytes32 messageDataHash,
+        uint256 tokenAmount
+    ) internal override returns (uint256) {
+        // fetch native token from sender if inbox doesn't already hold enough tokens to pay for fees
```
[the remark this answers is not in the document]
Comment (gvladika) [reply src/bridge/ERC20Inbox.sol]: done

File: src/bridge/ERC20Bridge.sol [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        // fetch native token from Inbox
+        IERC20(nativeToken).safeTransferFrom(msg.sender, address(this), amount);
+    }
+
+    function _executeLowLevelCall(
+        address to,
+        uint256 value,
+        bytes memory data
+    ) internal override returns (bool success, bytes memory returnData) {
+        if (to == nativeToken) {
```
[the remark this answers is not in the document]
Comment (gvladika) [reply src/bridge/ERC20Bridge.sol:55]: done

File: src/bridge/ERC20Bridge.sol [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            revert CallTargetNotAllowed(nativeToken);
+        }
+
+        // first release native token
+        IERC20(nativeToken).safeTransfer(to, value);
+        success = true;
```
[the remark this answers is not in the document]
Comment (gvladika) [reply src/bridge/ERC20Bridge.sol:61]: I checked using Foundry tests and it doesn't change gas consumption

File: src/bridge/Outbox.sol [hunk 6]
```diff
@@ -4,273 +4,16 @@
...
-        // we don't return the default context value to avoid a breaking change in the API
-        if (timestamp == TIMESTAMP_DEFAULT_CONTEXT) return uint256(0);
-        return uint256(timestamp);
-    }
-
-    /// @notice batch number is deprecated and now always returns 0
-    function l2ToL1BatchNum() external pure returns (uint256) {
+contract Outbox is AbsOutbox {
+    /// @inheritdoc AbsOutbox
+    function _defaultContextAmount() internal pure override returns (uint256) {
```
[the remark this answers is not in the document]
Comment (gvladika) [reply src/bridge/Outbox.sol:11]: done

File: src/rollup/AbsBridgeCreator.sol [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                        ""
+                    )
+                )
+            );
+            frame.outbox = Outbox(
+                address(new TransparentUpgradeableProxy(address(outboxTemplate), adminProxy, ""))
+            );
+        }
+
+        _initializeBridge(frame.bridge, IOwnable(rollup), nativeToken);
```
[the remark this answers is not in the document]
Comment (gvladika) [reply src/rollup/AbsBridgeCreator.sol]: Collapsed it back to single bridge creator contract

File: src/bridge/ERC20Outbox.sol [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+pragma solidity ^0.8.4;
+
+import "./AbsOutbox.sol";
+
+contract ERC20Outbox is AbsOutbox {
+    uint256 private constant AMOUNT_DEFAULT_CONTEXT = type(uint256).max;
+
+    function l2ToL1WithdrawalAmount() external view returns (uint256) {
+        uint256 amount = context.withdrawalAmount;
+        if (amount == AMOUNT_DEFAULT_CONTEXT) return 0;
```
[the remark this answers is not in the document]
Comment (gvladika) [reply src/bridge/ERC20Outbox.sol:15]: done, will also add to Notion docs and make sure it's part of officials docs later

**Review by yahgwai — APPROVED** [review]

**Review by gzeoneth — COMMENTED** [review]
some comment regarding upgradability, otherwise lgtm

File: src/bridge/AbsInbox.sol
[the remark this answers is not in the document]
Comment (gvladika) [reply src/bridge/AbsInbox.sol:1]: https://github.com/OffchainLabs/nitro-contracts/pull/19/commits/1198cc3b5b13a05978be6dc0db27805701bc8d94

File: src/bridge/AbsOutbox.sol
[the remark this answers is not in the document]
Comment (gvladika) [reply src/bridge/AbsOutbox.sol:1]: https://github.com/OffchainLabs/nitro-contracts/pull/19/commits/1198cc3b5b13a05978be6dc0db27805701bc8d94

File: src/bridge/AbsBridge.sol [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+    /// @dev get base fee which is emitted in `MessageDelivered` event and then picked up and
+    /// used in ArbOs to calculate the submission fee for retryable ticket
+    function _baseFeeToReport() internal view virtual returns (uint256);
+
+    /**
+     * @dev This empty reserved space is put in place to allow future versions to add new
+     * variables without shifting down storage in the inheritance chain.
+     * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+     */
+    uint256[50] private __gap;
```
[the remark this answers is not in the document]
Comment (gvladika) [reply src/bridge/AbsBridge.sol]: https://github.com/OffchainLabs/nitro-contracts/pull/19/commits/1198cc3b5b13a05978be6dc0db27805701bc8d94

File: package.json [hunk 10]
```diff
@@ -24,17 +28,24 @@
...
+    "test:e2e": "hardhat test test/e2e/*.ts",
     "postinstall": "patch-package",
     "deploy-factory": "hardhat run scripts/deployment.ts",
-    "deploy-rollup": "hardhat run scripts/rollupCreation.ts"
+    "deploy-eth-rollup": "hardhat run scripts/createEthRollup.ts",
+    "deploy-erc20-rollup": "hardhat run scripts/createERC20Rollup.ts"
   },
   "dependencies": {
+    "@arbitrum/sdk": "^3.1.3",
+    "@ethersproject/providers": "^5.7.2",
```
[the remark this answers is not in the document]
Comment (gvladika) [reply package.json]: https://github.com/OffchainLabs/nitro-contracts/pull/19/commits/60a1e5019e8a5c4d073f70f1714ab423969137ef

**Review by gzeoneth — APPROVED** [review]

**Pull request merged**